lib/AST/AutoDiff.cpp

```diff
@@ -85,3 +86,159 @@ unsigned autodiff::getOffsetForAutoDiffAssociatedFunction(
+///
+void AutoDiffParameterIndices::addAllNonSelfParams() {
+  indices.set(0, getNonSelfParamCount());
+}
+
+/// Adds the self parameter to the set. `isMethod` must be set. For example,
+///
+///   functionType = (Self) -> (A, B, C) -> R
+///   ==> adds "Self" to the set
+///
+void AutoDiffParameterIndices::addSelfParam() {
+  assert(isMethod &&
+         "trying to add self param to non-method parameter indices");
+  indices.set(indices.size() - 1);
+}
+
+/// Pushes the subset's parameter's types to `paramTypes`, in the order in
+/// which they appear in the function type. For example,
+///
+///   functionType = (A, B, C) -> R
+///   if "A" and "C" are in the set,
+///   ==> pushes {A, C} to `paramTypes`.
+///
+///   functionType = (Self) -> (A, B, C) -> R
+///   if "Self" and "C" are in the set,
+///   ==> pushes {Self, C} to `paramTypes`.
+///
+void AutoDiffParameterIndices::getSubsetParamTypes(
+    AnyFunctionType *functionType,
+    SmallVectorImpl<Type> &paramTypes) const {
+  AnyFunctionType *unwrapped = unwrapSelfParam(functionType, isMethod);
+  if (isMethod && indices[indices.size() - 1])
+    paramTypes.push_back(functionType->getParams()[0].getPlainType());
+  for (unsigned paramIndex : range(unwrapped->getNumParams()))
+    if (indices[paramIndex])
+      paramTypes.push_back(unwrapped->getParams()[paramIndex].getPlainType());
+}
+
+static unsigned countNumFlattenedElementTypes(Type type) {
+  if (auto *tupleTy = type->getCanonicalType()->getAs<TupleType>())
+    return accumulate(tupleTy->getElementTypes(), 0, [&](unsigned num, Type type) {
+      return num + countNumFlattenedElementTypes(type);
+    });
+  return 1;
+}
+
+/// Returns a bitvector for the SILFunction parameters corresponding to the
+/// parameters in this set. In particular, this explodes tuples and puts the
+/// method self parameter at the end. For example,
+///
+///   functionType = (A, B, C) -> R
+///   if "A" and "C" are in the set,
+///   ==> returns 101
+///   (because the lowered SIL type is (A, B, C) -> R)
+///
+///   functionType = (Self) -> (A, B, C) -> R
+///   if "Self" and "C" are in the set,
+///   ==> returns 0011
+///   (because the lowered SIL type is (A, B, C, Self) -> R)
+///
+///   functionType = (A, (B, C), D) -> R
+///   if "A" and "(B, C)" are in the set,
+///   ==> returns 1110
+///   (because the lowered SIL type is (A, B, C, D) -> R)
+///
+const llvm::SmallBitVector AutoDiffParameterIndices::getLowered(AnyFunctionType *functionType) const {
+  // Calculate the lowered sizes of all the parameters.
+  AnyFunctionType *unwrapped = unwrapSelfParam(functionType, isMethod);
+  SmallVector<unsigned, 8> paramLoweredSizes;
+  unsigned totalLoweredSize = 0;
+  auto addLoweredParamInfo = [&](Type type) {
+    unsigned paramLoweredSize = countNumFlattenedElementTypes(type);
+    paramLoweredSizes.push_back(paramLoweredSize);
+    totalLoweredSize += paramLoweredSize;
+  };
+  for (auto param : unwrapped->getParams())
```

Comment:
`AnyFunctionType::Param` is 3 words long. Pass by reference instead.
```suggestion
  for (auto &param : unwrapped->getParams())
```